python/ray/data/_internal/logical/rules/operator_fusion.py

```diff
@@ -0,0 +1,206 @@
+
+        # Allow fusing tasks->actors if the resources are compatible (read->map), but
+        # not the other way around. The latter (downstream op) will be used as the
+        # compute if fused.
+        if (
+            is_task_compute(down_logical_op._compute)
+            and isinstance(up_logical_op, AbstractMap)
+            and get_compute(up_logical_op._compute)
+            != get_compute(down_logical_op._compute)
+        ):
+            return False
+
+        # Fusing callable classes is only supported if they are the same function AND
+        # their construction arguments are the same.
+        # TODO(Clark): Support multiple callable classes instantiating in the same actor
+        # worker.
+        if (
+            isinstance(down_logical_op._fn, CallableClass)
+            and isinstance(up_logical_op, AbstractMap)
+            and isinstance(up_logical_op._fn, CallableClass)
+            and (
+                up_logical_op._fn != down_logical_op._fn
+                or (
+                    up_logical_op._fn_constructor_args
+                    != down_logical_op._fn_constructor_args
+                    or up_logical_op._fn_constructor_kwargs
+                    != down_logical_op._fn_constructor_kwargs
+                )
+            )
+        ):
+            return False
+
+        # Only fuse if the ops' remote arguments are compatible.
+        if not _are_remote_args_compatible(
+            up_logical_op._ray_remote_args or {}, down_logical_op._ray_remote_args or {}
+        ):
+            return False
+
+        # Otherwise, ops are compatible for fusion.
+        return True
+
+    def _fuse(self, down_op: PhysicalOperator, up_op: PhysicalOperator):
+        """Fuse the downstream operator with its upstream operator."""
+        from ray.data._internal.execution.operators.map_operator import MapOperator
+        from ray.data._internal.logical.operators.map_operator import AbstractMap
+
+        assert self._can_fuse(down_op, up_op)
+
+        # Fuse operator names.
+        name = up_op.name + "->" + down_op.name
+
+        down_logical_op = self._op_map.pop(down_op)
+        up_logical_op = self._op_map.pop(up_op)
+
+        # Merge target block sizes.
+        down_target_block_size = down_logical_op._target_block_size
+        up_target_block_size = (
+            up_logical_op._target_block_size
+            if isinstance(up_logical_op, AbstractMap)
+            else None
+        )
+        if down_target_block_size is not None and up_target_block_size is not None:
+            target_block_size = max(down_target_block_size, up_target_block_size)
+        elif up_target_block_size is not None:
+            target_block_size = up_target_block_size
+        else:
+            target_block_size = down_target_block_size
+
+        # Fuse transformation functions.
+        down_transform_fn = down_op.get_transformation_fn()
+        up_transform_fn = up_op.get_transformation_fn()
```

Comment:
let's assert `isinstance(down_transform_fn, MapTransformFn) and isinstance(up_transform_fn, MapTransformFn)`?

Reply:
Python doesn't support `isinstance` checks with subscripted generics.